### PR TITLE
Test postMessage() exception order

### PR DESCRIPTION
**Pull request by annevk**
Helps with https://github.com/whatwg/html/pull/3508.

### CODE REVIEW

**Comment by annevk** [comment]
It seems that Edge/Safari generally follow the specification here. Chrome/Firefox seem to handle the transfer errors later. I think either way we have to loop over the transferables before structured serializing to store them in memory, so there's probably no good reason to change the specification here.

**Comment by annevk** [comment]
Safari not throwing for duplicates: https://bugs.webkit.org/show_bug.cgi?id=183124.

**Comment by annevk** [comment]
Also:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1441141
* https://bugs.chromium.org/p/chromium/issues/detail?id=816447

**Comment by annevk** [comment]
And finally for Edge throwing the wrong exception for one test: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16104788/.

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/24881)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/346277974)

File: html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+function assert_transfer_error(transferList) {
+  assert_throws("DATA_CLONE_ERR", () => self.postMessage({ get whatever() { throw 1 } }, "*", transferList));
```
Comment (mkruisselbrink) [inline html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js]: nit: DataCloneError (at least http://web-platform-tests.org/writing-tests/testharness-api.html says the _ERR form is only supported "for compatibility with existing tests")

**Review by mkruisselbrink — APPROVED** [review]

**Comment by annevk** [comment]
I updated the tests to match the proposed change over at https://github.com/whatwg/html/pull/3557.

**Comment by annevk** [comment]
@jeremyroman could you perhaps review these tests? Suggestions for more tests?

File: html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+test(() => {
+  const b = new ArrayBuffer(1);
+  self.postMessage(null, "*", [b]);
+  assert_transfer_error([b], new Error("hi"));
+  assert_throws("DataCloneError", () => self.postMessage(null, '*', [b]));
+}, "Serialize throws before a transferred detached ArrayBuffer is found");
+
+promise_test(() => {
+  return self.createImageBitmap(document.createElement("canvas")).then(bitmap => {
```
Comment (jeremyroman) [inline html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js]: Do we want to be more exhaustive? We could do something like:

```js
function transfer_tests(name, create) {
  promise_test(async () => {
    const transferable = await create();
    assert_transfer_error([transferable, transferable]);
  }, `Cannot transfer the same ${name} twice`);

  promise_test(async () => {
    const transferable = await create();
    self.postMessage(null, "*", [transferable]);
    assert_throws("DataCloneError", () => self.postMessage(null, "*", [transferable]);
  }, `Serialize should make the ${name} detached, so it cannot be transferred again.`);

  promise_test(async () => {
    const transferable = await create();
    self.postMessage(null, "*", [transferable]);
    assert_transfer_error([transferable], new Error("hi"));
  }, `Serialize should throw before a detached ${name} is found.`);

  promise_test(async () => {
    const transferable = await create();
    let seen = false;
    const message = {
      get a() {
        self.postMessage(null, '*', [transferable]);
        seen = true;
      }
    };
    assert_throws("DataCloneError", () => self.postMessage(message, "*", [transferable]));
    assert_true(seen);
  }, `Cannot transfer ${name} detached while the message was serialized.`);
}

transfer_tests("ArrayBuffer", () => new ArrayBuffer(1));
transfer_tests("MessagePort", () => new MessageChannel().port1);
transfer_tests("ImageBitmap", () => self.createImageBitmap(document.createElement("canvas")));
// OffscreenCanvas? I don't believe we have any other transferables, at least implemented in Blink.
```

That might be excessively clever, but I think the logic for all of these tests generalizes to all kinds of transferable checks, so why not check the whole matrix?

File: html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+function assert_transfer_error(transferList, exception = "DataCloneError") {
+  assert_throws(exception, () => self.postMessage({ get whatever() { throw exception } }, "*", transferList));
```
Comment (jeremyroman) [inline html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js]: nit: Is this trying too hard to do what could be two separate functions? Throwing the string "DataCloneError" as the default behavior seems odd.

**Review by jeremyroman — COMMENTED** [review]
Makes sense to me overall.

File: html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js (hunk 2 above)
Comment (annevk) [reply html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js]: Yeah, why not. I'll make it so. Probably not today unfortunately.

File: html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js (hunk 3 above)
Comment (annevk) [reply html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js]: Will clean this up too.

**Comment by annevk** [comment]
Ended up making those fixes today anyway.

The one thing this loses is testing `ImageBitmap`'s `close()` method, but that should probably be tested as part of that object.

**Review by jeremyroman — APPROVED** [review]
LGTM; thanks!

**Pull request merged**